add test command definition (#886)
opentofu/opentofu: website/docs/cli/commands/test.mdx

---
description: >-
    The tofu test command performs integration tests of OpenTofu modules.
---

# Command: test

The `tofu test` command runs integration tests to ensure intended behavior of your OpenTofu configuration. It creates real
infrastructure, asserts conditional checks against the attributes of provisioned resources and attempts
to clean up the testing infrastructure upon completion.

## Usage

Usage: `tofu test [options]`

The `test` command executes the automated integration tests defined in `.tftest.hcl` files which OpenTofu will read
from the current configuration and testing directories.

OpenTofu will create real infrastructure according to the definition in the tested configuration. Then, it will
sequentially run the test blocks to perform conditional checks and assertions against attributes of provisioned resources.
Upon tests completion, OpenTofu will attempt to clean up created resources.

> _Note_: monitor the test output for successful cleanup.

Reviewer: We use `*Note*` in some places, and `_Note_` in others, let's be consistent

## General Options

> * `-test-directory=path` Set the test directory (default: "tests"). OpenTofu will search for test files in the set directory
> and in the directly of the test command execution.
> *Note* that the path shall be relative to the directory where `tofu test` is called.

Reviewer: ```suggestion
* `-test-directory=path` Set the test directory (default: "tests"). OpenTofu will search for test files in the set directory
as well as in the directly in which `tofu test` was executed.
*Note* that the path should be relative to the directory in which `tofu test` is called.
```

> * `-filter=testfile` Set the test file *path* within the `test-directory`. The option can be used multiple times to specify several files.

Reviewer: ```suggestion
* `-filter=testfile` Specify an individual test file *path* within the `test-directory` to use for testing. The option can be used multiple times to specify several files.
```

> * `-filter=testfile` Set the test file *path* within the `test-directory`. The option can be used multiple times to specify several files.
> *Note* that the path shall be relative to the directory where `tofu test` is called.

Reviewer: ```suggestion
* `-filter=testfile` Set the test file *path* within the `test-directory`. The option can be used multiple times to specify several files.
*Note* that the path should be relative to the directory in which `tofu test` is called.
```

* `-var 'foo=bar'` Set the input variables of the root module. The option can be used multiple times to set several variables.

* `-var-file=filename` Set the path to the file with variables values in addition to the default files terraform.tfvars and *.auto.tfvars.
The option can be used multiple times to specify several files.

* `-json` Set for test outputs to be formatted as JSON.

* `-no-color` Disable color codes in the command output.

* `-verbose` Print the plan or state for each test run block as it executes.

## Usage examples

> ### Flat structure

Reviewer 2: I feel like the docs now focus too much on the folder structure and about the combinations of `-filter` and `-test-directory`. It's probably enough to explain the interaction between them in a line or two, instead of creating a project structure mainly for that

Author: Hey Arel! Thanks for your feedback. The docu is WIP, however I wanted to share enough examples following GWT pattern as a response to this [feedback](https://github.com/opentofu/opentofu/issues/182). IMO if something is not intuitive, it shall be clearly illustrated to facilitate users onboarding. I feel that it's also influenced by the experience level of working with the testing feature. I'm trying to write docu assuming a reader who tries the tool for the first time.

│       ├── bar.tf
│       ├── bar.tftest.hcl
│       ├── main.tf
│       ├── main.tftest.hcl
│       ├── outputs.tf
│       └── variables.tf
├── outputs.tf
└── variables.tf
```

**WHEN** `tofu test` is executed in the project root directory,

**THEN** the tests defined in the following files will be run (note that the files are sorted in the execution order):

- bar.tftest.hcl,
- foo.tftest.hcl,
- main.tftest.hcl,
- tests/bar.tftest.hcl,
- tests/foo.tftest.hcl.

**WHEN** `tofu test -test-directory=modules/module_bar` is executed in the project root directory,

**THEN** the tests defined in the following files will be run (note that the files are sorted in the execution order):

- bar.tftest.hcl,
- foo.tftest.hcl,
- main.tftest.hcl,
- modules/module_bar/bar.tftest.hcl,
- modules/module_bar/foo.tftest.hcl,
- modules/module_bar/main.tftest.hcl.

**WHEN** `tofu test -test-directory=modules/module_bar -filter=bar.tftest.hcl` is executed in the project root directory,

**THEN** the tests defined in the file `bar.tftest.hcl` will be run only.

**WHEN** `tofu test -test-directory=modules/module_bar -filter=modules/module_bar/bar.tftest.hcl` is executed in the project root directory,

> **THEN** the tests defined in the file `modules/module_bar/bar.tftest.hcl` will be run only.

Reviewer: I understand we want to be comprehensive, but this is a very complex example, could we simplify this a bit more? I don't think we need foo/bar in every directory maybe?

Author: Thanks for idea - I adjusted the docu accordingly.

## Definition syntax

TBD

> ## Variables

Reviewer 2: Here I'd probably give an example of overriding variables in a test file / run block, and explain why one might want to do that

Author: Thanks for your feedback. I'll add a couple of illustrations/use cases. 

Could you please expand on this

> explain why one might want to do that

For context, I find that the main focus of this docu page shall be more practice oriented than explanation oriented (leaning more towards the top half of the [diagram](https://documentation.divio.com/) below).

![image](https://github.com/opentofu/opentofu/assets/13434797/d78ab91a-fb5f-468b-bfd7-0a6d9be1de50)

 WDYT?

Tests can be defined using variables specified in several ways.

> The variable values will be resolved according to the following order:

Reviewer: ```suggestion
Variable values will be resolved in the following order of precedence:
```

Author: Thanks for suggestion! It was implemented in line with this [subsection](https://opentofu.org/docs/language/values/variables/#variable-definition-precedence).

1. Environment variables with the prefix `TF_VAR_`;
2. tfvar files specified in terraform.tfvars, *.auto.tfvars;
3. Files specified using the flag `-var-file`;
4. Variables defined using the flag `-var`;
5. Variables from the `variable block` in test file;

> 6. Variables from the variable block in run block.

Reviewer 2: I think that we should already have docs on variable precedence. Maybe we can link there instead

Author: Could you please confirm if you mean this [subsection](http://localhost:3000/docs/language/values/variables/#variable-definition-precedence)? Thanks!

> _Note_ that the variable's value found at the later step takes precedence over the value at the earlier step.

Reviewer: This can be dropped if we go with the suggestion above.

---
description: >-
    The tofu test command performs integration tests of OpenTofu modules.
---

# Command: test

The `tofu test` command runs integration tests to ensure intended behavior of your OpenTofu configuration. It creates real
infrastructure, asserts conditional checks against the attributes of provisioned resources and attempts
to clean up the testing infrastructure upon completion.

## Usage

Usage: `tofu test [options]`

The `test` command executes the automated integration tests defined in `.tftest.hcl` files which OpenTofu will read
from the current configuration and testing directories.

OpenTofu will create real infrastructure according to the definition in the tested configuration. Then, it will
sequentially run the test blocks to perform conditional checks and assertions against attributes of provisioned resources.
Upon tests completion, OpenTofu will attempt to clean up created resources.

> _Note_: monitor the test output for successful cleanup.

Reviewer 3: Should this maybe be an info box instead of a Note?

Author: Thanks for suggestion Kuba! Implemented.

5. Report the test status.

:::info
Monitor the test output for successful cleanup.
:::

## General Options

* `-test-directory=path` Set the test directory (default: "tests"). OpenTofu will search for test files in the set directory
as well as in the directly in which `tofu test` was executed.

:::warning
The path should be relative to the directory in which `tofu test` is called.
:::

* `-filter=testfile` Specify an individual test file to use for testing.
The option can be used multiple times to specify several files.

:::warning
The path should be relative to the directory in which `tofu test` is called.
:::

* `-var 'foo=bar'` Set the input variables of the root module.
The option can be used multiple times to set several variables.

* `-var-file=filename` Set the path to the file with variables values in addition to the default
files terraform.tfvars and *.auto.tfvars. The option can be used multiple times to specify several files.

* `-json` Set for test outputs to be formatted as JSON.

* `-no-color` Disable color codes in the command output.

* `-verbose` Print the plan or state for each test run block as it executes.

## How to define a test suite

> The subsection describes how to filter test files when defining a test `suite`.

Reviewer: ```suggestion
In this subsection we will illustrate how to filter test files when defining a test `suite`.
```

### Flat structure

Assuming the OpenTofu module structure below

```commandline
.
├── LICENSE
├── README.md
├── main.tf
├── main.tftest.hcl
├── foo.tf
├── foo.tftest.hcl
├── bar.tf
├── bar.tftest.hcl
├── outputs.tf
└── variables.tf
```

- When you execute `tofu test` in the project directory, the test `suite` will consist of the files:
`bar.tftest.hcl`, `foo.tftest.hcl`, `main.tftest.hcl`.

> - To filter some files from the `suite`, use the `-filter` option.

Reviewer: ```suggestion
- To select some specific files from the `suite`, use the `-filter` option.
```

For example, when you execute `tofu test -test-directory=submodule -filter=foo.tftest.hcl` in the project directory,
no test files will be found.
:::

## How to define a test file

Tests are defined using three kinds of blocks in the way that every test file

- _shall_ contain _at least one_ `run` block to define a single test case;
- _may_ contain _at most one_ `variables` block to define "global" variables shared by all `runs` defined in a single file;
- _may_ contain _multiple_ `provider` blocks to configure providers shared by all `runs` in a single file.

:::info
Note that if a test file contains several `run` blocks, OpenTofu will execute them sequentially from the top to
the bottom of the file.
:::

### `run` block

<!-- details: /internal/configs/test_file.go:239 decodeTestRunBlock -->

The `run` block defines a _single test_. It specifies the operation which OpenTofu will execute,
sets mocks for providers,
and defines the checks to assess the state after the OpenTofu operation was executed.

The following non-mutually exclusive blocks and attributes can be defined within the scope of the `run` block to configure the test.

> | Name              | Type   | Multiple allowed | Description                                                                                                                          |

Reviewer: ```suggestion
| Name              | Type   | Multiple Allowed | Description                                                                                                                          |
```
